### PR TITLE
Added feature to have nested menu entries.

### PR DESCRIPTION
**Pull request by bangpradyumna**
Menu entries can now be nested upto any level. To add a sub-menu, Go to controllers.py of the respective folder and add a key 'entries' to the respective menu entry dictionary.

For example, To add a submenu as Config ---> Notif. Services --> My Sub Menu , do the following :

- Go to mod_config/controllers.py
- Change like 28,29 from  :
 `{'title': 'Notif. services', 'icon': 'bell-o', 'route': 'config.notifications'},`
 to :
  `{'title': 'Notif. services', 'icon': 'bell-o', 'route':
                'config.notifications' , 'entries' : [{'title': 'My Sub Entry', 'icon': 'my_icon', 'route':
                'my_route' }] },`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,4 +25,6 @@ secret_key
 config.py
 
 # Generated images for the honeypots (big binary blobs)
-honeypot_images/*.img
\ No newline at end of file
+honeypot_images/*.img
+/.idea/misc.xml
```
Comment (canihavesomecoffee) [inline .gitignore]: Three things I'd like to get clarification on for the .gitignore change:
- Why add these?
- Why the leading / ?
- They should be with the other project specific ignores

File: decorators.py [hunk 2]
```diff
@@ -1,8 +1,39 @@
+import copy
 from datetime import date
 from functools import wraps
+
 from flask import request, g, render_template
 
 
+def get_permissible_entries(user, entry):
```
Comment (canihavesomecoffee) [inline decorators.py:8]: I'd love to see some unittests that show that the code is working as intended. Any GSoC participant working on this project will have to do that anyway, so it's best to get used to it already :+1:

**Review by canihavesomecoffee — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (bangpradyumna) [reply .gitignore]: Those changes crept in due to my development environment and don't seem essential. I've removed them in latest commit.

File: decorators.py (hunk 2 above)
Comment (bangpradyumna) [reply decorators.py:8]: I'm working on it, Will update you soon :)

Comment (bangpradyumna) [reply decorators.py:8]: Unittests have been added in latest commits.

**Comment by canihavesomecoffee** [comment]
I'll check after the weekend, as it's too complext to check just as is :)

**Comment by bangpradyumna** [comment]
Alright 👍

**Comment by canihavesomecoffee** [comment]
@pep8speaks suggest diff

**Comment by bangpradyumna** [comment]
@pep8speaks pep8ify

File: test-menus.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    @patch('mod_auth.models.User')
+    def test_menu_with_partial_permissions(self, mock_user):
+        """
+        Passing a menu entry to the function get_menu_entries() when the user is not
+        allowed to access the route of one of the sub-entries with 'route' = 'honeypot.profiles'
+        This should return a dictionary without this entry.
+        """
+        mu = mock_user.return_value
+
+        def side_effect(*args):
```
Comment (canihavesomecoffee) [inline test-menus.py]: This method can be simplified...

`return args[0] != 'honeypot.profiles'`

File: test-menus.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import unittest
```
Comment (canihavesomecoffee) [inline test-menus.py]: Please rename the file. A dash in a filename is considered very bad practice.

File: test-menus.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    'config.data_processing'},
+                {'title': 'Honeypot services', 'icon': 'sliders', 'route':
+                    'config.services'}
+            ]
+        )
+        correct_entries = {'entries': [{'title': 'Notif. services', 'route': 'config.notifications', 'entries': [{'title': 'Notif. services', 'route': 'config.notifications', 'entries': [{'title': 'Honeypot services', 'route': 'config.services', 'icon': 'sliders'}], 'icon': 'bell-o'}], 'icon': 'bell-o'}, {'route': 'config.data_processing', 'title': 'Data processing', 'icon': 'exchange'}, {'route': 'config.services', 'title': 'Honeypot services', 'icon': 'sliders'}], 'icon': 'cog', 'title': 'Configuration'}
+        self.assertDictEqual(entries, correct_entries)
+
+
+if __name__=='__main__':
```
Comment (canihavesomecoffee) [inline test-menus.py]: This is fine as it's the only test right now, but this shouldn't be necessary when there's a more global setup of tests

File: test-menus.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import unittest
+from mock import patch
+from decorators import get_menu_entries
+
+class MenuTestCase(unittest.TestCase):
```
Comment (canihavesomecoffee) [inline test-menus.py]: These tests are good for the `get_menu_entries` method in general, but you should only test a single function with the tests. Now you test both `get_menu_entries` and `get_permissible_entries`. Add some separate tests for the latter and mock that function while testing the first.

File: test-menus.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                    'config.services', 'entries': [{'title': 'Profile mgmt', 'icon': 'bookmark', 'route':
+                    'honeypot.profiles'}]}]}]},
+                {'title': 'Data processing', 'icon': 'exchange', 'route':
+                    'config.data_processing'},
+                {'title': 'Honeypot services', 'icon': 'sliders', 'route':
+                    'config.services'}
+            ]
+        )
+
+        correct_entries={'entries': [{'title': 'Notif. services', 'route': 'config.notifications', 'entries': [{'title': 'Notif. services', 'route': 'config.notifications', 'entries': [{'title': 'Honeypot services', 'route': 'config.services', 'entries': [{'title': 'Profile mgmt', 'route': 'honeypot.profiles', 'icon': 'bookmark'}], 'icon': 'sliders'}], 'icon': 'bell-o'}], 'icon': 'bell-o'}, {'route': 'config.data_processing', 'title': 'Data processing', 'icon': 'exchange'}, {'route': 'config.services', 'title': 'Honeypot services', 'icon': 'sliders'}], 'icon': 'cog', 'title': 'Configuration'}
```
Comment (canihavesomecoffee) [inline test-menus.py]: Try to reformat this over multiple lines so it's easier to read :+1:

**Review by canihavesomecoffee — COMMENTED** [review]
Good start, but more work is needed.

Also try to enable PEP-8 support in your IDE, so you can get hints/warnings. It'll have to be correct in order to be merged.

File: test-menus.py (hunk 5 above)
Comment (bangpradyumna) [reply test-menus.py]: Are you talking about line 93 or line 90 ?

File: test-menus.py (hunk 6 above)
Comment (bangpradyumna) [reply test-menus.py]: Working on it :+1:

File: test-menus.py (hunk 5 above)
Comment (canihavesomecoffee) [reply test-menus.py]: 93. The entire section to run it standalone basically :)

File: test-menus.py (hunk 3 above)
Comment (bangpradyumna) [reply test-menus.py]: Thanks for the input, Done :+1:

File: test-menus.py (hunk 5 above)
Comment (bangpradyumna) [reply test-menus.py]: Done in the latest commits.

File: test-menus.py (hunk 7 above)
Comment (bangpradyumna) [reply test-menus.py]: Done in the latest commit.

File: test-menus.py (hunk 4 above)
Comment (bangpradyumna) [reply test-menus.py]: Done.

**Comment by bangpradyumna** [comment]
@pep8speaks suggest diff

**Comment by bangpradyumna** [comment]
@canihavesomecoffee The changes suggested by pep8tify seem to be IDE specific customizations?
I enabled pep8 on pycharm and this is what it gave me after expanding that one line dictionary to multiple lines. What do you suggest?

**Comment by bangpradyumna** [comment]
@pep8speaks suggest diff

**Comment by bangpradyumna** [comment]
@canihavesomecoffee Seems like the nested dictionary indentation differs from IDE to IDE as is the case in above diff of pep8tify
I have made the major changes asked by you.
In the unittests I import functions as `server.decorators.get_menu_entries()` and I expect that any new user renames the directory from `/usr/src/pipot/pipot-server` to `/usr/src/pipot/server` (as mentioned in the install instructions)
Do you want me to change install.sh such that it renames root directory from `pipot-server` to `server`

**Review by canihavesomecoffee — APPROVED** [review]

**Pull request merged**